compiler/qsc_frontend/src/resolve.rs

```diff
@@ -101,18 +101,22 @@ pub(super) enum Error {
     Unimplemented(String, #[label] Span),
 }
 
-struct Scope {
+#[derive(Debug, Clone)]
+pub struct Scope {
+    span: Span,
     kind: ScopeKind,
     opens: FxHashMap<Rc<str>, Vec<Open>>,
     tys: FxHashMap<Rc<str>, ItemId>,
     terms: FxHashMap<Rc<str>, ItemId>,
-    vars: FxHashMap<Rc<str>, NodeId>,
+    // u32 is the `valid_at` - the lowest offset at which the variable name can be used
+    vars: FxHashMap<Rc<str>, (u32, NodeId)>,
```

Comment:
would reversing this data structure help? Using `FxHashMap<NodeId, (u32, Rc<str>)>` would make sure that shadowed variables aren't overwritten but rather are all present.

Reply:
Oh, never mind. I keep forgetting which stage of compilation we are at, and here we need to match on string because it's during resolution, not afterward.